bundles/core/org.eclipse.smarthome.core.thing/src/main/java/org/eclipse/smarthome/core/thing/internal/i18n/ChannelI18nUtilImpl.java

```diff
@@ -0,0 +1,104 @@
+/**
+ * Copyright (c) 2014,2018 Contributors to the Eclipse Foundation
+ *
+ * See the NOTICE file(s) distributed with this work for additional
+ * information regarding copyright ownership.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ */
+package org.eclipse.smarthome.core.thing.internal.i18n;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Locale;
+import java.util.function.Function;
+
+import org.eclipse.jdt.annotation.NonNullByDefault;
+import org.eclipse.jdt.annotation.Nullable;
+import org.eclipse.smarthome.core.thing.i18n.ChannelGroupTypeI18nLocalizationService;
+import org.eclipse.smarthome.core.thing.i18n.ChannelI18nUtil;
+import org.eclipse.smarthome.core.thing.i18n.ChannelTypeI18nLocalizationService;
+import org.eclipse.smarthome.core.thing.i18n.ThingTypeI18nLocalizationService;
+import org.eclipse.smarthome.core.thing.type.ChannelDefinition;
+import org.eclipse.smarthome.core.thing.type.ChannelType;
+import org.eclipse.smarthome.core.thing.type.ChannelTypeRegistry;
+import org.eclipse.smarthome.core.thing.type.ChannelTypeUID;
+import org.osgi.framework.Bundle;
+import org.osgi.service.component.annotations.Component;
+import org.osgi.service.component.annotations.Reference;
+
+/**
+ * A utility service which localises {@link ChannelDefinition}.
+ * Falls back to a localised {@link ChannelType} for label and description when not given otherwise.
+ *
+ * @see {@link ThingTypeI18nLocalizationService}
+ * @see {@link ChannelGroupTypeI18nLocalizationService}
+ *
+ * @author Henning Treu - initial contribution
+ *
+ */
+@NonNullByDefault
+@Component(immediate = true, service = ChannelI18nUtil.class)
+public class ChannelI18nUtilImpl implements ChannelI18nUtil {
```

Comment:
This one strucks me weird. We have so far `XXXI18nUtil` classes, which are no DS components and are simply instantiated by passing a `TranslationProvider`.

Doing an interface here with an impl class, which is then registered as a component seems to fall out of this pattern.
Is this rather supposed to be a `ChannelI18nLocalizationService` (or even a `ChannelDefinitionI18nLocalizationService`) instead?

Reply:
Yes, the first idea was to just expose the implementation class as a service. I had an offline review and this was one of the comments. will change back.

Reply:
I'm afraid, you didn't read all of my comment:

> We have so far XXXI18nUtil classes, which are no DS components and are simply instantiated by passing a `TranslationProvider`.

You have now refactored the class by removing the interface, but it still is a DS component.